### PR TITLE
APPS/IO-DEMO: Handle SIGINT signal to gracefully terminate client/server

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Handle SIGINT signal to gracefully terminate client/server.

## Why ?

This is needed to close all Network objects (e.g. UCX worker, EPs, etc) and destroy memory for all allocated operationss/callbacks.

## How ?

1. Move `_status` field to `P2pDemoCommon` class which is common for `DemoClient`/`DemoServer`.
2. Set signal actions from `vfs_server_sighandler()` to be able raise the signal to the old signal handler.
3. Handle `SIGINT` signal in IO-Demo signal handler and set `_status` to `RUNTIME_EXCEEDED`.
4. Stop server process by checking `_status != OK`.

### CODE REVIEW

File: src/tools/vfs/vfs_server.c [hunk 1]
```diff
@@ -32,6 +32,7 @@ typedef struct {
 } vfs_serever_fd_state_t;
 
 typedef struct {
+    struct sigaction       old_sigact;
```
Comment (yosefe) [inline src/tools/vfs/vfs_server.c:35]: why do we care about prev sighandler?

File: test/apps/iodemo/io_demo.cc [hunk 2]
```diff
@@ -670,6 +678,17 @@ class P2pDemoCommon : public UcxContext {
         MemoryPool<SendCompleteCallback>& _pool;
     };
 
+    static void signal_terminate_handler(int signo)
+    {
+        _status = RUNTIME_EXCEEDED;
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:683]: maybe add new enum TERMINATE_SIGNALED ?

File: src/tools/vfs/vfs_server.c (hunk 1 above)
Comment (dmitrygx) [reply src/tools/vfs/vfs_server.c:35]: otherwise, IO-demo signal handler isn't called

Comment (yosefe) [reply src/tools/vfs/vfs_server.c:35]: do we have another signal handler?

Comment (dmitrygx) [reply src/tools/vfs/vfs_server.c:35]: btw, it also could be problematic for user's application which set the same signals which are handled by VFS.

File: test/apps/iodemo/io_demo.cc (hunk 2 above)
Comment (dmitrygx) [reply test/apps/iodemo/io_demo.cc]: done

File: src/tools/vfs/vfs_server.c (hunk 1 above)
Comment (yosefe) [reply src/tools/vfs/vfs_server.c:35]: vfs_server is an application , not part of libucs
it should not affect io_demo
also io_demo is an application so no other application could exist when it's running

Comment (dmitrygx) [reply src/tools/vfs/vfs_server.c:35]: I thought that VFS is a part of UCX library.
Reverted code that takes care of old sigactions.

Comment (yosefe) [reply src/tools/vfs/vfs_server.c]: > I thought that VFS is a part of UCX library.
> Reverted code that takes care of old sigactions.

Some VFS common infra is part of libucs, but the daemon (ucx_vfs mount tool) is standalone, like ucx_perftest

File: test/apps/iodemo/io_demo.cc [hunk 3]
```diff
@@ -16,6 +16,8 @@
 #include <unistd.h>
 #include <cstdlib>
 #include <ctime>
+#include <csignal>
+#include <cerrno>
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:20]: still needed?

Comment (dmitrygx) [reply test/apps/iodemo/io_demo.cc:20]: yes, we are printing error if `sigaction` faield

Comment (yosefe) [reply test/apps/iodemo/io_demo.cc:20]: ok

**Review by yosefe — APPROVED** [review]

**Pull request merged**